src/main/java/net/dries007/tfc/client/overworld/LevelRendererExtension.java

```diff
@@ -555,58 +556,61 @@ public boolean tickRain(ClientLevel level, int ticks, Camera camera)
                     {
-                        break;
-                    }
+                        cursor.move(Direction.DOWN);
+                        addedAnyParticles = true;
+
+                        // Don't check the biome, as all overworld biomes should support rain, and we checked that above
 
-                    final double offsetX = random.nextDouble();
-                    final double offsetZ = random.nextDouble();
-                    final BlockState state = level.getBlockState(cursor);
-                    final FluidState fluid = level.getFluidState(cursor);
-
-                    // Handle `IBlockRain`, which needs to pretend the block is a solid block.
-                    final VoxelShape shape = state.getBlock() instanceof IBlockRain
-                        ? Shapes.block()
-                        : state.getCollisionShape(level, cursor);
-                    final double offsetY = Math.max(
-                        shape.max(Direction.Axis.Y, offsetX, offsetZ),
-                        fluid.getHeight(level, cursor)
-                    );
-
-                    final ParticleOptions options = !fluid.is(FluidTags.LAVA)
-                        && !state.is(Blocks.MAGMA_BLOCK)
-                        && !CampfireBlock.isLitCampfire(state)
-                        ? ParticleTypes.RAIN
-                        : ParticleTypes.SMOKE;
-                    level.addParticle(options, cursor.getX() + offsetX, cursor.getY() + offsetY, cursor.getZ() + offsetZ, 0.0, 0.0, 0.0);
+                        // With minimal particle options, we still search for a position (as we use it to determine if we need to play rain sounds), but we
+                        // stop at the first one, before spawning any particles for it.
+                        if (minecraft.options.particles().get() == ParticleStatus.MINIMAL)
+                        {
+                            break;
+                        }
+
+                        final double offsetX = random.nextDouble();
+                        final double offsetZ = random.nextDouble();
+                        final BlockState state = level.getBlockState(cursor);
+                        final FluidState fluid = level.getFluidState(cursor);
+
+                        // Handle `IBlockRain`, which needs to pretend the block is a solid block.
+                        final VoxelShape shape = state.getBlock() instanceof IBlockRain
+                            ? Shapes.block()
+                            : state.getCollisionShape(level, cursor);
+                        final double offsetY = Math.max(
+                            shape.max(Direction.Axis.Y, offsetX, offsetZ),
+                            fluid.getHeight(level, cursor)
+                        );
+
+                        final ParticleOptions options = !fluid.is(FluidTags.LAVA)
+                            && !state.is(Blocks.MAGMA_BLOCK)
+                            && !CampfireBlock.isLitCampfire(state)
```

Comment:
we have our own magma blocks, this is missing them already. need something more sturdy.

Reply:
Added a tag for blocks that smoke in rain, and added magma blocks to it. Also made our firepits and forges smoke in rain.